sql/core/src/main/scala/org/apache/spark/sql/functions.scala

```diff
@@ -652,6 +652,122 @@ object functions {
+   * is the relative error of the approximation.
+   *
+   * @group agg_funcs
+   * @since 3.0.0
+   */
+  def percentile_approx(columnName: String, percentage: Array[Double], accuracy: Long): Column = {
+    percentile_approx(Column(columnName), percentage, accuracy)
+  }
+
+  /**
+   * Aggregate function: Returns and array of the approximate percentile values
+   * of numeric column col at the given percentages.
+   *
+   * Each value of the percentage array must be between 0.0 and 1.0.
+   *
+   * The accuracy parameter is a positive numeric literal
+   * which controls approximation accuracy at the cost of memory.
+   * Higher value of accuracy yields better accuracy, 1.0/accuracy
+   * is the relative error of the approximation.
+   *
+   * @group agg_funcs
+   * @since 3.0.0
+   */
+  def percentile_approx(e: Column, percentage: Seq[Double], accuracy: Long): Column = {
+    percentile_approx(e, percentage.toArray, accuracy)
+  }
+
+  /**
+   * Aggregate function: Returns and array of the approximate percentile values
+   * of numeric column col at the given percentages.
+   *
+   * Each value of the percentage array must be between 0.0 and 1.0.
+   *
+   * The accuracy parameter is a positive numeric literal
+   * which controls approximation accuracy at the cost of memory.
+   * Higher value of accuracy yields better accuracy, 1.0/accuracy
+   * is the relative error of the approximation.
+   *
+   * @group agg_funcs
+   * @since 3.0.0
+   */
+  def percentile_approx(columnName: String, percentage: Seq[Double], accuracy: Long): Column = {
+    percentile_approx(Column(columnName), percentage.toArray, accuracy)
+  }
+
+  /**
+   * Aggregate function: Returns the approximate percentile value of numeric
+   * column col at the given percentage.
+   *
+   * The value of percentage must be between 0.0 and 1.0.\
+   *
+   * The accuracy parameter is a positive numeric literal
+   * which controls approximation accuracy at the cost of memory.
+   * Higher value of accuracy yields better accuracy, 1.0/accuracy
+   * is the relative error of the approximation.
+   *
+   * @group agg_funcs
+   * @since 3.0.0
+   */
+  def percentile_approx(e: Column, percentage: Double, accuracy: Long): Column = {
+    withAggregateFunction {
+      new ApproximatePercentile(
+        e.expr, lit(percentage).expr, lit(accuracy).expr
+      )
+    }
+  }
+
+  /**
+   * Aggregate function: Returns the approximate percentile value of numeric
+   * column col at the given percentage.
+   *
+   * The value of percentage must be between 0.0 and 1.0.\
+   *
+   * The accuracy parameter is a positive numeric literal
+   * which controls approximation accuracy at the cost of memory.
+   * Higher value of accuracy yields better accuracy, 1.0/accuracy
+   * is the relative error of the approximation.
+   *
+   * @group agg_funcs
+   * @since 3.0.0
+   */
+  def percentile_approx(columnName: String, percentage: Double, accuracy: Long): Column = {
```

Comment:
@zero323, I think it's better to have only one function that takes columns. See:

https://github.com/apache/spark/blob/1f50a5875b46885a40668c058a1a28e736776244/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L57-L59

Reply:
Do you mean we should remove variants `(String, _, _) -> Column` and leave only `(Column, _, _) -> Column`? Or rather replace all variants with `(Column, Column, Column) -> Column`?

I'll be happy to do the former, but the latter is probably a bad idea:

- Aggregation by group gets confusing. 
- Array variants get somewhat painful to use for anyone but Scala users ‒ building array literal in Java is not fun. Same about non-JVM languages (and if we allow for non `Column` arguments then, complexity of wrappers grows).

Reply:
This is what I meant `(Column, Column, Column) -> Column` to have a single method to cover all other cases.
What we need to do might be just wrap it by `lit` or `col: `percentile_approx(col(...), lit(...), lit(...))` so won't be too difficult. We don't need to duplicate docs with less maintenance.

I did quick test and seems working fine in general:

Scala:

```scala
scala> spark.range(1).select(lit(Array(1, 2, 3))).show()
```
```
+---------+
|  [1,2,3]|
+---------+
|[1, 2, 3]|
+---------+
```

Java:

```java
spark.range(1).select(lit(new String[]{"a", "b"})).show();
```
```
+------+
| [a,b]|
+------+
|[a, b]|
+------+
```

Python:

```python
>>> from pyspark.sql.functions import lit
>>> import array
>>> spark.range(1).select(lit(["a", "b"]))
```

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../spark/python/pyspark/sql/functions.py", line 54, in _
    jc = getattr(sc._jvm.functions, name)(col._jc if isinstance(col, Column) else col)
  File "/.../spark/python/lib/py4j-0.10.8.1-src.zip/py4j/java_gateway.py", line 1286, in __call__
  File "/.../spark/python/pyspark/sql/utils.py", line 98, in deco
    return f(*a, **kw)
  File "/.../spark/python/lib/py4j-0.10.8.1-src.zip/py4j/protocol.py", line 328, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling z:org.apache.spark.sql.functions.lit.
: java.lang.RuntimeException: Unsupported literal type class java.util.ArrayList [a, b]
	at org.apache.spark.sql.catalyst.expressions.Literal$.apply(literals.scala:85)
	at org.apache.spark.sql.catalyst.expressions.Literal$.$anonfun$create$2(literals.scala:145)
	at scala.util.Failure.getOrElse(Try.scala:222)
	at org.apache.spark.sql.catalyst.expressions.Literal$.create(literals.scala:145)
	at org.apache.spark.sql.functions$.typedLit(functions.scala:131)
	at org.apache.spark.sql.functions$.lit(functions.scala:114)
	at org.apache.spark.sql.functions.lit(functions.scala)
```

R:

```R
> collect(select(createDataFrame(mtcars), lit(list('a', 'b', 'c'))))
   [a,b,c]
1  a, b, c
2  a, b, c
```

Seems we should fix PySpark's `lit` to take `list` as an array. It won't be too difficult - we can check if this is a list with numeric elements, and create an Array manually.

Reply:
To be honest I am not very enthusiastic about it and I am not even convinced that it is consistent with the rest of `functions`. 

The closest equivalents we have are

- `approx_count_distinct` with `rsd`
- `last` with `ignoreNulls`

and both use external types, not columns. Not to mention this is still counter-intuitive and painful to use though:

> we don't need to duplicate docs with less maintenance.

is a fair point.

- I can easily remove `Seq` variants, that's for sure and cut number of signatures by two, leaving us with four.
- If having not `Column` variant on JVM is fine, we can drop `(String, _, _) => Column` variant so that brings us to two variants.
- It is also not hard to build `Column` objects transparently for Python and R users to support `(Column, Column, Column) => Column`. But I am still concerned about confusing semantics. 

  If two variants are still to much, we could always have `(Column, Any, Double) => Column` ‒ `o.a.sql.functions` is already quite full of `Any`s.